packages/connect-core/src/transform-iterable-story.spec.ts

```diff
@@ -33,6 +33,78 @@ import {
   transformParse,
 } from "./transform-iterable.js";
 
+class Reader<T> {
+  private it: AsyncIterable<T>;
+
+  constructor(it: AsyncIterable<T>) {
+    this.it = it;
+  }
+  [Symbol.asyncIterator](): AsyncIterator<T> {
+    return {
+      next: async () => {
+        for await (const item of this.it) {
+          return { value: item };
+        }
+        return { value: "", done: true };
+      },
+    };
+  }
+}
+
+class Writer<T> {
+  private queue: IteratorResult<T>[] = [];
+  // Represents the resolve function of the promise returned by the async iterator if no values exist in the queue at
+  // the time of request.  It is resolved when a value is successfully received into the queue.
+  private queueResolver: ((val: IteratorResult<T>) => void) | undefined;
+  // Represents the resolve function of the promise returned by send.  It is resolved when a value is successfully
+  // read via the async iterator.
+  private sendResolver: (() => void) | undefined;
+
+  async process(item: IteratorResult<T, undefined>) {
+    // If there is an iterator resolver then a consumer of the async iterator is waiting on a value.  So resolve that
+    // promise with the new value being sent and return a promise that is immediately resolved
+    if (this.queueResolver) {
+      this.queueResolver(item);
+      this.queueResolver = undefined;
+      return new Promise<void>((resolve) => {
+        resolve();
+      });
+    }
+    // Otherwise no one is waiting on a value yet so add it to the queue and return a promise that will be resolved
+    // when someone reads this value
+    this.queue.push(item);
+    return new Promise<void>((resolve) => {
+      this.sendResolver = resolve;
+    });
+  }
+  async send(item: T) {
```

Comment:
One thing we don't account for yet is errors. Consider a client doing this:

```ts
async function f(w: Writer) {
  const dbCon = ...
  try {
    for (;;) {
      const row = dbConn.getNextRow();
      if (!row) break;
      await w.write( row );
    }
  } finally {
    dbConn.close();
  }
}
```

We need some way to tell the calling code to stop in case of an error. Our `CallOptions` also accept a `signal?: AbortSignal` that needs to stop the calling code when triggered (AbortError).

One option is to implement `throw` in the async iterator. The the code handling this `Writer` needs to make use of it. Another option is to have a facility like `abort(reason)` in `Writer`. 

There is actually a similar situation for handlers, where we want to abort the asynchronous iterable of a server-streaming RPC implementation in case of an error (request aborted, error when serializing the response, or the client just hung up). We do not have a solution for this case in handlers yet either.

Reply:
@timostamm I pushed up some changes that took a stab at trying to handle these errors using a `throw` implementation.  It _seems_ ok, although I'm not confident it's correct and probably needs work.  Please take a look and let me know.

One thing I'm stuck on is that I'd like the consumer to be able to call `throw` on the `Reader` and have that bubble up to the `Writer` also by calling its `throw`.  But the `pipe` call returns a generator function, which behaves differently when `throw` is called (it throws the error given).

Reply:
Using AsyncIterator.throw sounds great. From the tests that https://github.com/bufbuild/connect-web/pull/431, it seems like a decent approach.

Do we need the capability to throw on the reader? Clients and handlers will operate on the iterable, and we only have [StreamingConn.read](https://github.com/bufbuild/connect-web/blob/3b8d028a88c509a92f841fd3c262273bcbbaa4f6/packages/connect-core/src/interceptor.ts#L273) reading from an asynchronous iterable. I think it should be fine to have users use an AbortSignal if they want to abort a request because they can't process response messages?